solar_consumer/save/save_data_platform.py

```diff
@@ -0,0 +1,191 @@
+
+"""
+
+import os
+from loguru import logger
+
+from dp_sdk.ocf import dp
+from grpclib.client import Channel
+import pandas as pd
+
+
+data_platform_host = os.getenv("DATA_PLATFORM_HOST", "localhost")
+data_platform_port = int(os.getenv("DATA_PLATFORM_PORT", "50051"))
+
+
+async def save_to_generation_to_data_platform(
+    data_df: pd.DataFrame, client: dp.DataPlatformDataServiceStub | None = None
+):
+    """
+    Save Generation data to the Data-platform.
+
+    0. First we get all the locations
+
+    Here's how we do it for each gsp
+    1. Get all the locations
+    2. Create an observer for that regime if it doesn't already exist
+    3. For each gsp: Get only the data for that gsp
+    4. Get the location for that gsp
+    5. Get the start and end timestamps from that data
+    6. Get the most recent observations for that location and observer,
+    7. Remove any data points from our data that are already in the database
+    8. Update location capacity based on the max capacity in this data
+    9. Create new observations for the remaining data points
+
+    :param data_df: DataFrame containing forecast data with required columns.
+    """
+
+    assert "target_datetime_utc" in data_df.columns
+    assert "solar_generation_kw" in data_df.columns
+    assert "gsp_id" in data_df.columns
+    assert "regime" in data_df.columns
+    assert "capacity_mwp" in data_df.columns
+
+    # Initialize the Data Platform client, if not there already provided
+    if client is None:
+        channel = Channel(host=data_platform_host, port=data_platform_port)
+        client = dp.DataPlatformDataServiceStub(channel)
+
+    # 1. Get all locations
+    all_locations = await get_all_gsp_and_national_locations(client)
+
+    # 2. Create an observer for that gsp and regime if it doesn't already exist
+    regime = data_df["regime"].unique()
+    assert len(regime) == 1, "DataFrame must contain only one regime type"
+    regime = regime[0]
+    name = f"PVLive-consumer-{regime}".lower()
+    observer_request = dp.CreateObserverRequest(name=name)
+    try:
+        _ = await client.create_observer(observer_request)
+    except Exception:
+        # TODO Future: Get Observer, and if try to make one.
+        logger.warning(f"Observer {name} probably already exists, so carrying on anyway.")
+
+    # for each gsp
+    gsp_ids = data_df["gsp_id"].unique()
+    for gsp_id in gsp_ids:
+        logger.info(f"Saving GSP ID: {gsp_id} to Data Platform")
+
+        # 3. Get only the data for that gsp
+        gsp_data = data_df[data_df["gsp_id"] == gsp_id]
+
+        # 4. Get the location for that gsp
+        location = all_locations.get(gsp_id)
+        if location is None:
+            logger.warning(f"No location found for GSP ID {gsp_id}, skipping.")
+            continue
+        location_uuid = location.location_uuid
+
+        # 5. Get the start and end timestamps from that data
+        # TODO Future: remove 5,6,7 if database can deal with duplicates
```

Comment:
I'm doing this at the moment - I think we shouldn't merge this in until that is done and this change is made here.